crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move

```diff
@@ -0,0 +1,1045 @@
+        dkg_output: vector<u8>,
+        hashed_message: vector<u8>,
+        presign: vector<u8>,
+        centralized_signed_message: vector<u8>,
+    }
+
+    /// Event emitted to start a batched sign process.
+    ///
+    /// ### Fields
+    /// - **`session_id`**: The session identifier for the batched sign process.
+    /// - **`hashed_messages`**: A list of hashed messages to be signed.
+    /// - **`initiator`**: The address of the user who initiated the protocol.
+    public struct StartBatchedSignEvent has copy, drop {
+        session_id: ID,
+        hashed_messages: vector<vector<u8>>,
+        initiator: address
+    }
+
+    /// Event emitted to start a batched presign flow, i.e. a flow that creates multiple presigns at once.
+    ///
+    /// ### Fields
+    /// - **`session_id`**: The session identifier for the batched sign process.
+    /// - **`batch_size`**: The number of presign sessions to be started.
+    public struct StartBatchedPresignEvent has copy, drop {
+        session_id: ID,
+        batch_size: u64,
+        initiator: address
+    }
+
+    // <<<<<<<<<<<<<<<<<<<<<<<< Error codes <<<<<<<<<<<<<<<<<<<<<<<<
+    /// Error raised when the sender is not the system address.
+    const ENotSystemAddress: u64 = 0;
+    const EMessageApprovalDWalletMismatch: u64 = 1;
+    const EDwalletMismatch: u64 = 2;
+    const EApprovalsAndMessagesLenMismatch: u64 = 3;
+    const EMissingApprovalOrWrongApprovalOrder: u64 = 4;
+    const ECentralizedSignedMessagesAndMessagesLenMismatch: u64 = 5;
+    const EPresignsAndMessagesLenMismatch: u64 = 6;
+    const EInvalidSignatures: u64 = 7;
+    const EApprovalsAndSignaturesLenMismatch: u64 = 8;
+    // >>>>>>>>>>>>>>>>>>>>>>>> Error codes >>>>>>>>>>>>>>>>>>>>>>>>
+
+    // <<<<<<<<<<<<<<<<<<<<<<<< Constants <<<<<<<<<<<<<<<<<<<<<<<<
+    /// System address for asserting system-level actions.
+    const SYSTEM_ADDRESS: address = @0x0;
+    // >>>>>>>>>>>>>>>>>>>>>>>> Constants >>>>>>>>>>>>>>>>>>>>>>>>
+
+    /// Starts the first Distributed Key Generation (DKG) session.
+    ///
+    /// This function creates a new `DWalletCap` object,
+    /// transfers it to the session initiator,
+    /// and emits a `StartDKGFirstRoundEvent` to signal
+    /// the beginning of the DKG process.
+    ///
+    /// ### Effects
+    /// - Generates a new `DWalletCap` object.
+    /// - Transfers the `DWalletCap` to the session initiator (`ctx.sender`).
+    /// - Emits a `StartDKGFirstRoundEvent`.
+    ///
+    /// ### Emits
+    /// - `StartDKGFirstRoundEvent`:
+    ///   - `session_id`: The generated session ID.
+    ///   - `initiator`: The address of the transaction sender.
+    ///   - `dwallet_cap_id`: The ID of the created `DWalletCap`.
+    public fun launch_dkg_first_round(ctx: &mut TxContext) {
+        let dwallet_cap = create_dwallet_cap(ctx);
+        let dwallet_cap_id = object::id(&dwallet_cap);
+        transfer::public_transfer(dwallet_cap, tx_context::sender(ctx));
```

Comment:
Are we sure the `dwallet_cap` should be created here, before the DKG MPC session even begun? The user will have access to a `dwallet_cap` with no valid `DWallet` in the system after you call `public_transfer`. @omersadika is this the right design? What if the DKG session never completes, or has an error etc.?